nmcli.py by itself

alcamie101 opened the pull request:
If you could review this thanks Will

Commit:
nmcli.py by itself

willthames commented on library/nmcli.py:
```diff
@@ -0,0 +1,735 @@
+options:
+    name:
+        required: true
+        version_added: "0.1"
```
version_added relates to Ansible version. And only really needed if it differs to the version_added of the module itself.

willthames commented on library/nmcli.py:
```diff
@@ -0,0 +1,735 @@
+author: Chris Long
+version_added: "0.1"
+short_description: Manage Networking
+requirements: [ conadd, condel, conmod, conup, condown, constatus, conshow ]
```
what are these?

alcamie101 replied:
thanks for pointing that out. Should be just NMCLI at present

willthames commented on library/nmcli.py:
```diff
@@ -0,0 +1,735 @@
+# To change the property of a setting e.g. MTU, issue a command as follows:
+- nmcli: name=modify cname=my-eth1 mtu=9000
+
+    Exit Status's:
```
Do we care about the Exit Statuses here?

alcamie101 replied:
Just notes to assist with debugging.

willthames commented on library/nmcli.py:
```diff
@@ -0,0 +1,735 @@
+        self.mac=module.params['mac']
+        self.vlanid=module.params['vlanid']
+        # The following is going to be used in dbus code
+        self.devtypes={1: "Ethernet",
```
I'd make a constant called DEVTYPES and refer to that here.

willthames commented on library/nmcli.py:
```diff
@@ -0,0 +1,735 @@
+                       14: "Generic",
+                       15: "Team"
+                       }
+        self.states={0: "Unknown",
```
same for constant dict STATES

willthames commented on library/nmcli.py:
```diff
@@ -0,0 +1,735 @@
+        # we are going to use name and type in this instance to find if that connection exists and is of type x
+        connections=self.list_connection_info()
+
+        for con_item in connections:
```
return con_item in connections

willthames commented on library/nmcli.py:
```diff
@@ -0,0 +1,735 @@
+        supports_check_mode=True
+    )
+
+    # if module.params['state'] is None and module.params['enabled'] is None:
```
delete these, and set them to required=True

willthames commented on library/nmcli.py:
```diff
@@ -0,0 +1,735 @@
+        #     result['Connection']='Yessss. Connection %s exists' % nmcli.cname
+        #     result['Connection']='Connection %s is being set to down' % nmcli.cname
+        if rc!=0:
+            module.fail_json(name =('No Connection named %s exists' % nmcli.cname), msg=err, rc=rc)
```
Don't fail when a connection that is specified as absent is already absent. exit with changed=False.

willthames commented on library/nmcli.py:
```diff
@@ -0,0 +1,735 @@
+    result['state']=nmcli.state
+
+    if nmcli.state=='absent':
+        # if nmcli.connection_exists():
```
Does connection_exists() work - I'd think this should be the priority,

alcamie101 replied:
I was getting some weird happenings the other day. connection_exists() does now work. Will put that test back in.

willthames commented on library/nmcli.py:
```diff
@@ -0,0 +1,735 @@
+
+    def modify_connection(self):
+        cmd=[]
+        if self.type=='team':
```
you can create a dict of type to method call

commands = { 'team': self.modify_connection_team, 'team-slave', ... }
return self.execute_command(commands[self.type])

willthames commented on library/nmcli.py:
```diff
@@ -0,0 +1,735 @@
+    module=AnsibleModule(
+        argument_spec=dict(
+            enabled=dict(required=False, default=None, choices=['yes', 'no'], type='str'),
+            name=dict(required=False, default=None, choices=['add', 'mod', 'show', 'up', 'down', 'del'], type='str'),
```
this is a bad field name. Use 'action' or similar. Also, suggest having nice option names like 'modify' and 'delete' and converting while required. 

Actually, this field shouldn't even exist. Use enabled and state to cover these cases (you can add 'show' to state)

Commit:
Added create_team_slave, modify_team_slave, up_connection and rename
connection_down to up_connection

willthames commented on library/nmcli.py:
```diff
@@ -0,0 +1,735 @@
+        argument_spec=dict(
+            enabled=dict(required=False, default=None, choices=['yes', 'no'], type='str'),
+            name=dict(required=False, default=None, choices=['add', 'mod', 'show', 'up', 'down', 'del'], type='str'),
+            state=dict(default=None, choices=['present', 'absent', 'modify'], type='str'),
```
modify is not a state. Use present, and check if the parameters are the same, modifying if necessary

alcamie101 replied:
Yeh. Cutting down the number of fields required is definitely a good thing

willthames commented on library/nmcli.py:
```diff
@@ -0,0 +1,735 @@
+        cmd.append('type')
+        cmd.append(self.type)
+        cmd.append('con-name')
+        if self.cname is not None:
```
this whole next section should just be replaced with some kind of for loop

```
for attr in ['cname', 'ip4, 'gw4', ... ]: 
    if getattr(self, attr) is not None: 
        cmd.append(attr)
        cmd.append(getattr(self, attr))
```

alcamie101 replied:
OK I'll try that.

willthames commented on library/nmcli.py:
```diff
@@ -0,0 +1,735 @@
+        elif self.ifname is not None:
+            cmd.append(self.ifname)
+        else:
+            self.module.fail_json(msg="You haven't specified a name for the connection")
```
Handle this case long before it gets to this method call. This method shouldn't need to access module at all.

alcamie101 replied:
Yeh, that way there are a lot less fail_json's to write

Commit:
The code is not yet pretty but team and team-slave are working
Idempotency needs alot more var checks to get it to work, that's the
next thing on TODO list.

Commit:
added support for Bondind and Ethernet

Commit:
corrected bugs and added vars to support Vlans and bridging
